tinygrad/codegen/uopgraph.py

```diff
@@ -385,15 +387,33 @@ def loop_collapse(loop_start, loop_end, compval, idx, mval, multconst, rng):
   (UOp.max(UOp.cvar('c'), -(UOp(UOps.RANGE).name('s'))), lambda c,s: -s if -(s.src[1].arg-1) >= c.arg else None),
   (UOp.max(UOp.cvar('c'), -(UOp(UOps.RANGE).name('s')+UOp.cvar('c2'))), lambda c,s,c2: -(s+c2) if -(s.src[1].arg-1+c2.arg) >= c.arg else None),
   # const rules
+  # TODO: Write a UPat for vec(8) types
   (UOp(UOps.GEP, src=(UOp.cvar("c"),)).name("root"), lambda root, c: UOp.const(root.dtype, c.arg)),
+  (UPat(UOps.GEP, name="root", src=(UPat(UOps.CONST, name="c"),)), lambda root, c: UOp.const(root.dtype, c.arg)),
+  (UPat(UOps.GEP, name="root", src=(UPat(UOps.VECTORIZE, src=(
+    UPat(UOps.CONST, name="c1"), UPat(UOps.CONST, name="c2"))))), lambda root, c1, c2: UOp.const(root.dtype, c1.arg if root.arg == 0 else c2.arg)),
+  (UPat(UOps.GEP, name="root", src=(UPat(UOps.VECTORIZE, src=(
+    UPat(UOps.CONST, name="c1"), UPat(UOps.CONST, name="c2"), UPat(UOps.CONST, name="c3"), UPat(UOps.CONST, name="c4"))))),
+    lambda root, c1, c2, c3, c4: UOp.const(root.dtype, [c1.arg, c2.arg, c3.arg, c4.arg][root.arg])),
   (UPat(UOps.CAST, name="root", src=UPat(UOps.CONST, name="c")), lambda root, c: UOp.const(root.dtype, c.arg)),
-  (UPat(UOps.VECTORIZE, name="root", src=UPat(UOps.CONST, name="c")), lambda root, c: UOp.const(root.dtype, c.arg)),
-  # a phi on a DEFINE_ACC without loops or a CONST is a noop. this is for correctness, not just speed
+  # a phi on a DEFINE_ACC without loops, CONST, or a vectorized CONST is a noop. this is for correctness, not just speed
   (UOp(UOps.PHI, src=(UOp(UOps.DEFINE_ACC).name("acc"), UOp.var("acc"))), lambda acc: UOp.cast(acc.src[0], acc.dtype)),
   (UOp(UOps.PHI, src=(UOp(UOps.DEFINE_ACC, src=(UOp.cvar(),)), UOp.var("x"))), lambda x: x),
+  (UPat(UOps.PHI, src=(UPat(UOps.DEFINE_ACC, src=(UPat(UOps.VECTORIZE, src=[UPat(UOps.CONST) for _ in range(2)]))), UPat(name="x"))), lambda x: x),
+  (UPat(UOps.PHI, src=(UPat(UOps.DEFINE_ACC, src=(UPat(UOps.VECTORIZE, src=[UPat(UOps.CONST) for _ in range(4)]))), UPat(name="x"))), lambda x: x),
+  (UPat(UOps.PHI, src=(UPat(UOps.DEFINE_ACC, src=(UPat(UOps.VECTORIZE, src=[UPat(UOps.CONST) for _ in range(8)]))), UPat(name="x"))), lambda x: x),
   (UOp(UOps.PHI, src=(UOp.cvar(), UOp.var("x"))), lambda x: x),
+  (UPat(UOps.PHI, src=(UPat(UOps.VECTORIZE, src=[UPat(UOps.CONST) for _ in range(2)]), UPat(name="x"))), lambda x: x),
+  (UPat(UOps.PHI, src=(UPat(UOps.VECTORIZE, src=[UPat(UOps.CONST) for _ in range(4)]), UPat(name="x"))), lambda x: x),
+  (UPat(UOps.PHI, src=(UPat(UOps.VECTORIZE, src=[UPat(UOps.CONST) for _ in range(8)]), UPat(name="x"))), lambda x: x),
```

Comment:
I'm supportive of a well-tested PR for generic vectorize fold, we repeat this here too https://github.com/tinygrad/tinygrad/blob/0b3a34e3b1a7b8f73432d3ee4d555911c5cb80ab/tinygrad/codegen/uopgraph.py#L339-L341

Reply:
By generic, did you mean creating a function that does this processing? I spent some time thinking about how I could do this with a function, but struggled to come up with a good design that covers all required cases succinctly. I chose to generate these rules with a list comprehension and unpack them inline instead. I feel that it's sufficiently concise, but more flexible.

Reply:
Added some tests for these rules, too.